packages/artifact_proxy/lib/src/artifact_proxy.dart

```diff
@@ -0,0 +1,86 @@
+// ignore_for_file: avoid_print
+
+import 'package:artifact_proxy/artifact_proxy.dart';
+import 'package:collection/collection.dart';
+import 'package:shelf/shelf.dart';
+
+/// A [Handler] that proxies artifact requests to the correct location.
+/// This is determined based on the [config].
+Handler artifactProxyHandler({required ProxyConfig config}) {
+  final engineMappings = config.engineMappings;
+  final shorebirdEngineRevisions = engineMappings.keys.cast<String>();
+
+  return (Request request) {
+    final path = request.url.path;
+    final shorebirdEngineRevision = shorebirdEngineRevisions.firstWhereOrNull(
+      path.contains,
+    );
+
+    final normalizedPath = shorebirdEngineRevision != null
+        ? path.replaceAll(shorebirdEngineRevision, r'$engine')
+        : path;
+
+    if (shorebirdEngineRevision == null) {
+      final location = getFlutterArtifactLocation(artifactPath: normalizedPath);
+      print('No engine revision detected, forwarding to: $location');
```

Comment:
Definitely not blocking on this PR, but should we eventually move to using some kind of logger class? (probably https://pub.dev/packages/mason_logger)